fix: Make toJson and fromJson always compatible

klkucaj opened the pull request:
Closes: #2149 

`toJson` now always creates a JSON of the entire model resulting on consistent roundtrips between `toJson` and `fromJson`. 

In order to introduce this change, a rework was made to the Serverpod Framework's serialization of server models.

Key changes in this PR:
-  We've introduced a new interface called `ProtocolSerialization` which includes a method named `toJsonForProtocol`. The interface and method is used by the Serverpod framework for protocol communication. Meaning that only fields indented for the client are included when called on Serverpod Models. The interface can also be implemented by custom classes that want to restrict their serialization through the Serverpod Framework.
- Renamed the `allToJson` method to `toJson`. This new `toJson` method serializes all fields, including those not serialized by the `toJsonForProtocol` method.
- We have also introduced a new static method to the `SerializationManager`, named `encodeForProtocol`. Now, whenever you need to serialize a model for protocol communication, you should use `SerializationManager.encodeForProtocol` instead of the `SerializationManager.encode` method.

It's important to note that the old `toJson` method only serialized fields for protocol communication, whereas the new `toJson` method will serialize all fields.

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

## Breaking changes:
- Remove the `allToJson()` method on Server models.
- The `toJson()` now returns a JSON of the entire model.
- Introduce the `ProtocolSerialization` interface with the method `toJsonForProcotol()` that is used by the framework to create serializations of Server models that are sent to the client.

SandPod commented on packages/serverpod_serialization/lib/src/serialization.dart:
```diff
@@ -13,21 +13,25 @@ typedef constructor<T> = T Function(
+/// The [ConstraintSerialization] defines a toJson method which makes it
+/// possible to limit what fields are serialized
+abstract interface class ConstraintSerialization {
+  /// Returns a serialized JSON structure of the model, ready to be sent
+  /// through the API. This does not include fields that are marked as
+  /// database only.
```
Suggestion: this description does not need to talk about what fields are included or not. Just mention what the serialization is used for.

SandPod commented on tests/serverpod_test_server/test/generated_methods/all_to_json_test.dart:
```diff
@@ -10,7 +10,7 @@ void main() {
     var types = Types();
 
-    var jsonMap = types.allToJson();
+    var jsonMap = types.toJson();
```
Suggestion: Rename test file to match the `toJson()` name.

SandPod commented on tests/serverpod_test_server/test/generated_methods/to_json_test.dart:
```diff
@@ -621,7 +621,7 @@ void main() {
     );
 
-    var jsonMap = object.toJson();
+    var jsonMap = object.toConstraintJson();
```
Suggestion: The tests in this file should probably be moved to the `all_to_json_test.dart` file and then this file should contain tests that validate the behavior of the `toConstraintJson()` method.

SandPod replied:
Please make sure to validate all test cases are needed and are adapted to the new names.

SandPod replied:
We probably want a couple of more `toContraintJson()` scenarios as well that validate the different cases we can have. For example a List containing a class with server only fields.

SandPod commented on tools/serverpod_cli/lib/src/analyzer/models/validation/restrictions.dart:
```diff
@@ -93,7 +93,7 @@ const _databaseModelReservedFieldNames = [
   'find',
   'setColumn',
-  'allToJson',
+  'toJson',
```
chore: Change this to `toContraintJson`/`toProtocolJson` name.

SandPod replied:
`toJson` is already present in the globallyRestrictedKeywords.

SandPod commented on tools/serverpod_cli/lib/src/generator/dart/library_generators/model_library_generator.dart:
```diff
@@ -155,9 +155,16 @@ class SerializableModelLibraryGenerator {
       } else {
         classBuilder.extend =
             refer('SerializableEntity', serverpodUrl(serverCode));
+        if (serverCode) {
```
Suggestion: Move the if check and the contents out to above line 147.

There is no need to nest this branching and repeat the code of implementing the interface.

SandPod commented on tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/field_restricted_names_test.dart:
```diff
@@ -94,7 +94,7 @@ void main() {
     'find',
     'setColumn',
-    'allToJson',
+    'toJson',
```
Chore: update this in the same way as I commented on the restricted field implementation.

SandPod commented on tools/serverpod_cli/test/generator/dart/client_code_generator/class_field_test.dart:
```diff
@@ -72,6 +72,16 @@ void main() {
             reason: 'Missing extends clause for SerializableModel.');
       });
 
+      test('does not implement ConstraintSerialization.', () {
```
Awesome ⭐️

SandPod commented on packages/serverpod_serialization/lib/src/serialization.dart:
```diff
@@ -13,21 +13,25 @@ typedef constructor<T> = T Function(
+  /// Returns a serialized JSON structure of the model, ready to be sent
+  /// through the API. This does not include fields that are marked as
+  /// database only.
+  dynamic toConstraintJson();
```
Suggestion: Use `Map<String, dynamic>` as return type.

Is there any reason for not having that return type here directly? We code generate it into the classes.

If we do it, would it make sense to make the same change for the other `toJson` method?

klkucaj replied:
for 'toJsonForProtocol' we can change the return type, but for 'toJson' we cant because 'SerializableEntity' is extended by other types like Enum, and when an enum overrides 'toJson' it returns 'int' or 'String'.
So this change can only be done for 'toJsonForProtocol'.

SandPod commented on packages/serverpod_serialization/lib/src/serialization.dart:
```diff
@@ -13,21 +13,25 @@ typedef constructor<T> = T Function(
+
 /// Get the type provided as an generic. Useful for getting a nullable type.
 Type getType<T>() => T;
 
```
Chore: Use the `toConstraintJson` for serialization.

There is one major thing missing with this PR that we didn't catch in testing.

We never actually use the `toConstraintJson` method for serialization.

We need to add tests to ensure we catch this in the future.
We also need to ensure that this flow works with custom classes. I don't think just changing how we serialize on line 169 in this file would fix that. We also need to make sure that we can serialize a custom class that is a field of another class.

SandPod commented on tests/serverpod_test_server/lib/src/endpoints/json_protocol.dart:
```diff
@@ -0,0 +1,36 @@
+import 'package:serverpod_test_server/src/generated/protocol.dart';
+
+class JsonProtocolEndpoint extends Endpoint {
+  Future<ScopeServerOnlyField> getJsonForProtocol(
```
Suggestion(NIT): Rename to `getScopeServerOnlyField`.

This better describes what we are getting. The `jsonForProtocol` is an implementation detail that doesn't need to be reflected in the method.

SandPod commented on tests/serverpod_test_server/lib/src/endpoints/json_protocol.dart:
```diff
@@ -0,0 +1,36 @@
+  Future<ScopeServerOnlyField> getJsonForProtocol(
+    Session session,
+  ) async {
+    return ScopeServerOnlyField(
```
Suggestion: Add a value to `serverOnlyScope` on the first level as well and validate that we don't pass that in the body.

Same for the streaming case.

SandPod commented on tests/serverpod_test_server/test_e2e/json_protocol_test.dart:
```diff
@@ -0,0 +1,118 @@
+      );
+    });
+
+    test('then the nested object should not contain server-only field', () {
+      Map jsonMap = jsonDecode(response.body);
+      Map nestedMap = jsonMap['nested'];
+      expect(
+        nestedMap,
+        isNot(contains('serverOnlyScope')),
+      );
+    });
+  });
```
Suggestion: Add first level validation that the field is not included.

As suggested in the endpoint, add a field for the first level as well and validate that it is not inlcuded.

If I look at the model, the `nested` field holds the same model again. To validate that the first level server only field is not included we should validate that `jsonMap['serverOnlyScope'];` does not exist.

It is good to validate the nested case as well.

SandPod replied:
Make sure to validate the "first-level" and "nested" object for a streaming connection as well.

SandPod commented on tests/serverpod_test_server/test_e2e/json_protocol_test.dart:
```diff
@@ -0,0 +1,118 @@
+      },
+    );
+
+    test(
+      'then the serialized response JSON body should contain "object" key',
+      () async {
+        Map responseMap = jsonDecode(message);
+        expect(responseMap, contains('object'));
+      },
+    );
```
Suggestion: Please add validation for all fields included in the body. Such as `endpoint` and `className`.

These tests are really important since they will help us keep the protocol consistent.

SandPod commented on tests/serverpod_test_server/test/generated_methods/to_json_test.dart:
```diff
@@ -611,8 +611,7 @@ void main() {
 
-  test(
-      'Given an object with a server only field then that field is not serialized.',
+  test('Given an object with a server only field then field is serialized.',
```
Suggestion: This is actually a nested field, please reflect that in the description.

It would be good to add a field for the first level to ensure we serialize them.

SandPod commented on tests/serverpod_test_server/test/generated_methods/to_json_for_protocol.dart:
```diff
@@ -611,7 +611,8 @@ void main() {
   });
 
-  test('Given an object with a server only field then field is serialized.',
+  test(
+      'Given an object with a server only field then that field is not serialized.',
       () {
```
Question: Doesn't it make more sense to use this file for the ´toJson` tests now and then use the other file as the `toJsonForProtocol()` tests?

That way we would have the same behavior for all tests here and we could prove that it is just a method call change but the `allToJson()` behavior is now preserved in the `toJson()` method while the `toJson()` behavior is now preserved in the `toJsonForProtocol()` method.

SandPod reviewed (commented):
Almost there, a couple of comments and then we should be done.

SandPod reviewed (approved):
LGTM! 🚀 

Great work on this! 🙌 
This is a big improvement to our serialization.

Could you add a note in the description around the changes made for "encode" and how that influences for example our Redis cache so that we now actually cache the full object?

klkucaj commented:
Key changes in this PR:
- Renamed the `toJson` method to `toJsonForProtocol`. This method serializes fields specifically suited for protocol communication.
- Renamed the `allToJson` method to `toJson`. This new `toJson` method serializes all fields, including those not serialized by the `toJsonForProtocol` method.
- Removed `toJsonForProtocol` from the `SerializableEntity`. It is now part of the `ProtocolSerialization` interface and will only be implemented by models that require the `toJsonForProtocol` method.

It's important to note that the old `toJson` method only serialized fields for protocol communication, whereas the new `toJson` method will serialize all fields.

SandPod reviewed (changes requested):
CustomClasses are not implemented!

SandPod commented on tests/serverpod_test_server/lib/src/protocol_custom_classes.dart:
```diff
@@ -0,0 +1,38 @@
+// ignore: depend_on_referenced_packages
+import 'package:serverpod_client/serverpod_client.dart';
+
+class ProtocolCustomClass extends SerializableEntity
```
Suggestion: Remove extention on `SerializableEntity`, this is not required for a custom class and we don't need to validate that in this test.

SandPod replied:
Same for the class implementation on the client side.

klkucaj replied:
In order to remove SerializableEntity from custom classes we need to update serverpod streaming, otherwise it is not possible to do this because all models/classes are required to extend SerializableEntity for serverpod streaming.

SandPod reviewed (changes requested):
Just remove the extension on SerializablyEntity then we are ready.

SandPod reviewed (approved):
Now we are ready. 🎉
